### ISSUE TITLE
Virtual Devices Take a Long Time to Respond

### ISSUE THREAD

**Issue by markaslett** [Colorado4Wheeler/HomeKit-Bridge#88]
I recently added a virtual device which consists of multiple switches to HomeKit Bridge
when I operate the switch, it works but it takes a while (spinning circle in top right of icon) and the Indigo Log shows the error message below:

 Action Group                    Family Room Lights On
   Virtual Devices                 sent "Family Room Lights" on
   Action Group                    Family Room Front Lamp On
   Virtual Devices                 sent "Family Room Front Lamp" on
   Action Group                    Family Room Rear Lamp On
   Virtual Devices                 sent "Family Room Rear Lamp" on
   Sent INSTEON                    "Family Room Main Lights" on to 100
   Sent INSTEON                    "Front Reading Lamp" on to 100
   Sent INSTEON                    "Rear Reading Lamp" on to 75

Mar 27, 2018 at 7:20:32 PM
   HomeKit Bridge Error            Maximum time exceeded while setting the 'On' HomeKit characteristic for 'Family Room Lights' (HomeKit device 'Family Room Lights'), aborting attempt.  This can happen if you try to set a device to a state is is already in (i.e., turning off a device that is already off).

**Comment by Colorado4Wheeler** [comment]
A virtual device GROUP or a virtual ON/OFF device?  Check whichever it is in Indigo, does is properly indicate that is is on or off when you operate the member devices?  Considering a switch is quite easily handled in HomeKit this sounds like to me that the on state is not getting updated, which would cause a beachball because the plugin is waiting for confirmation that the change requested has been made to the device and if on state doesn’t properly and promptly represent a change then you’ll get the timeout error.  Virtual groups can be touchy to set up and often can stay on all of the time or don’t turn off and if either of these things happen then you are going to get a beachball.

**Comment by markaslett** [comment]
Also, when I click on the switch again the devices do not switch off

**Comment by markaslett** [comment]
Its a virtual on off device that operates an action group made up of individual lights

**Comment by Colorado4Wheeler** [comment]
And in the device list in Indigo, when you click “Turn On” or “Turn Off” can you see that the state of the device in question on the Indigo list is toggling appropriately and instantly?

**Comment by markaslett** [comment]
<img width="1440" alt="screen shot 2018-03-27 at 7 35 22 pm" src="https://user-images.githubusercontent.com/36767367/38000555-177a6358-31f6-11e8-9183-f1db1733ff9b.png">

**Comment by Colorado4Wheeler** [comment]
So your problem is that you are not reporting status and if you don’t then the device doesn’t actually turn on or off and we need that to happen.  So you will need to check the box “supports status” and figure out what indicates that this is on or off.  You’ll notice when you toggle this that the Indigo device list doesn’t change, it stays off all of the time and this is why you are beach balling.

**Comment by markaslett** [comment]
The lights turn on immediately. The server pauses with the spinning ball the shows a timeout. The turn off action does not work

**Comment by markaslett** [comment]
The supports status option would need a variable which doesn't make sense for a regular switch

**Comment by Colorado4Wheeler** [comment]
Yes, it does, but it’s not a regular switch in Indigo, it’s a “half switch” in Indigo if it’s not reporting an On or Off state.  I may write something specifically for this in the future but the way to make it work right this moment is to define a status.  Without that status how is HomeKit supposed to know if it is on or off or if the command to turn it on or turn it off was successful?  Think about it from HomeKit’s perspective, you tap the switch to turn it on, the plugin does turn it on but because there’s not status when HomeKit asks “so, are you on” our answer is “I don’t know”.

**Comment by markaslett** [comment]
So if I use a virtual device group and switch on status it works as opposed to using a virtual on/off device that controls an on and off action groups that does not work. 

I have a lot of virtual switches that I will need to convert. I use them to control my sonos, TVs, Harmony Hubs etc via Alexa. They're very useful.

**Comment by Colorado4Wheeler** [comment]
Virtual switches work just fine, I use them myself too, they just need to get a **status** from a variable, that's it.  A virtual switch is comprised of action groups to turn on and turn off, right?  So, in that action group add an action that sets a variable to `true` or `false` and then reference that variable.

For one of my TV's I have a virtual switched called "TV", it executes an action group to send an IR command for on and other for off.  I simply created a variable called 'tvIsOn' and when I run the IR group to turn it on I tacked on an extra action that toggles that variable On, the opposite when I run the other group and then in the virtual group I reference "tvIsOn' as my on/off state and it all works beautifully.  A very simple solution, you don't need to scrap anything at all and that may take 15 minutes of your life to add that functionality to your actions.

**Comment by Colorado4Wheeler** [comment]
So my setup is this (a little more complex because my TV doesn't have discreet power reporting)
![screen shot 2018-03-27 at 5 58 41 pm](https://user-images.githubusercontent.com/20329385/38001175-a9b0bf14-31e8-11e8-84e9-509020c0bad4.png)
![screen shot 2018-03-27 at 5 58 48 pm](https://user-images.githubusercontent.com/20329385/38001177-acd41a7e-31e8-11e8-9eb9-b9e806724ada.png)
![screen shot 2018-03-27 at 5 59 09 pm](https://user-images.githubusercontent.com/20329385/38001178-b2289338-31e8-11e8-8fa0-386fb8a82dc0.png)

In your case you could simply use the built-in variable tools:

![screen shot 2018-03-27 at 5 59 25 pm](https://user-images.githubusercontent.com/20329385/38001197-c139149c-31e8-11e8-97d6-cb61deab5774.png)

**Comment by markaslett** [comment]
Thanks I understand I have used them quite a bit. 

I have 10 sonos's each with 15-20 virtual devices to control volume, channel selection etc and 7 TV's each with 10 virtual devices as well as at least one virtual lighting device group per room. Put another way I will need to create a lot of variables to track everything. That said its possible. 

Other than the required changes everything has worked great. I was following the closed beta and the back and forth on the UI. I have to say you have done a great job. Its super easy to set up and very intuitive. Not easy todo - well done and thanks.  

I have a very extensive indigo setup and look forward to continuing to experiment with your plugin.  Thanks, Mark

**Comment by Colorado4Wheeler** [comment]
Alternatively you can wait, I've flagged this as an enhancement so I'll add it to the queue of things to support in a way that will support status or no status.  You could also create a switch of the switch that serves the current needs and reports status until the feature is added - a bit redundant but it's only purpose would be to report on and toggle the more complicated switches you have.

**Comment by markaslett** [comment]
<img width="1440" alt="screen shot 2018-03-27 at 8 13 39 pm" src="https://user-images.githubusercontent.com/36767367/38001554-68c7ff86-31fb-11e8-91d3-de359bb500a5.png">

**Comment by Colorado4Wheeler** [comment]
You might want to wait anyway, it's on my priority list to build a **system** for Sonos (and iTunes) into this plugin, with play controls, volume controls, playlist reporting and the works.

But, yeah, that's a few virtual devices :).

**Comment by markaslett** [comment]
ok I probably will. Once again thank you and congrats great body of work.
